lib/utils.js

```diff
@@ -64,54 +64,37 @@ function parseCapsForInnerDriver (jsonwpCaps, w3cCapabilities, constraints={}, d
 
   // Get MJSONWP caps
   let desiredCaps = {};
+  let processedJsonwpCaps = null;
   if (hasJSONWPCaps) {
     desiredCaps = jsonwpCaps;
+    processedJsonwpCaps = {...desiredCaps};
   }
 
   // Get W3C caps
+  let processedW3CCapabilities = null;
   if (hasW3CCaps) {
     // Call the process capabilities algorithm to find matching caps on the W3C
     // (see: https://github.com/jlipps/simple-wd-spec#processing-capabilities)
     let matchingW3CCaps = processCapabilities(w3cCapabilities, constraints, true);
+    desiredCaps = matchingW3CCaps;
 
-    if (hasJSONWPCaps) {
-      let jsonwpKeys = _.keys(desiredCaps);
-      let matchingW3CCapsKeys = _.keys(matchingW3CCaps);
+    // Create a new w3c capabilities payload that contains only the matching caps in `alwaysMatch`
+    processedW3CCapabilities = {
+      alwaysMatch: {...insertAppiumPrefixes(desiredCaps)},
+      firstMatch: [{}],
+    };
 
-      // If JSONWP has keys that W3C doesn't have, log a warning message
-      let differingKeys = _.differenceWith(jsonwpKeys, matchingW3CCapsKeys);
+    // If we found extraneuous keys in JSONWP caps, fall back to JSONWP
+    if (hasJSONWPCaps) {
+      let differingKeys = _.differenceWith(_.keys(jsonwpCaps), _.keys(matchingW3CCaps));
       if (!_.isEmpty(differingKeys)) {
-        logger.warn(`The following capabilities were provided in the JSONWP desired capabilities but
-          not the W3C capabilities: ${JSON.stringify(differingKeys)}`);
+        logger.warn(`The following capabilities were provided in the JSONWP desired capabilities that are missing
```

Comment:
this will look weird in the log with a newline; i would just use concatenation to put the string across lines